### PR TITLE
Harden metrics, lifecycle handling, and CLI validation

### PR DESCRIPTION
**Pull request by tonyredondo**
## Summary

This PR hardens TimeItSharp across configuration validation, scenario execution, lifecycle callbacks, runtime metrics, exporters, CLI behavior, packaging, and CI.

### Configuration and CLI

- Added strict configuration validation with aggregated, actionable errors for counts, timing/statistical settings, scenarios, process data, metrics, and extension entries.
- Validate configurations before cloning or starting any process, including malformed JSON and missing configuration files.
- Improved extension loading for in-memory, named, and file-based extensions with descriptive failures.
- Reworked command parsing to preserve quoted executable paths and arguments, including paths containing spaces.
- Fixed `key=value` variable parsing so values containing additional `=` characters and empty values are preserved.
- Made explicit CLI options override JSON settings without unintentionally replacing omitted values; `--metrics false` is supported.
- Return reliable non-zero exit codes without terminating the process before cleanup callbacks run.

### Lifecycle and scenario robustness

- Guarantee scenario cleanup, `AfterAllScenariosFinishes`, and `OnFinish` callbacks run when initialization, execution, exporter, or callback handling fails.
- Continue invoking later lifecycle subscribers after an earlier subscriber throws, while preserving the first failure.
- Correct failed-result timestamps and ensure failed/path-validation scenarios trigger scenario-finish callbacks.
- Isolate `PATH` changes per child process instead of mutating the host environment.
- Prevent environment-variable collection mutation during template expansion.
- Use a `TimeSpan` duration budget so sub-minute execution time is not truncated.
- Handle invalid working directories, malformed metrics files, and empty/failed result sets without unrelated crashes.

### Metrics and exporters

- Corrected user/system/total CPU metric ordering.
- Added protection against overlapping metric pushes and made startup-hook initialization/process-exit handling idempotent and shutdown-safe.
- Hardened distribution charts against duplicate names, empty/non-finite data, constant distributions, and invalid bin counts.
- Made JSON export create parent directories, truncate stale files, preserve in-memory results, and redact sensitive environment values.
- Made Datadog export lazy, resilient to empty/failed scenarios, status-aware, and safe around sensitive environment metadata.
- Fixed profiler environment variable names, profiler asset discovery, and native-loader validation.

### Packaging, tests, and CI

- Added 25 xUnit regression tests covering validation, builders, template variables, lifecycle callbacks, exporters, and utility/statistical edge cases.
- Added `InternalsVisibleTo` for focused engine coverage.
- Replaced the incomplete direct tracer dependency with `Datadog.Trace.Bundle` 3.43.0 while retaining `Datadog.Trace.BenchmarkDotNet` 2.61.0 compatibility.
- Added direct `Spectre.Console` packaging metadata and the Common package README.
- Pinned SDK 10.0.300 with `global.json` and expanded CI across net6.0 through net10.0.
- Added vulnerability auditing, trimmed sample publishing, package creation, and local tool-install verification to CI.

## Validation performed

- `dotnet restore`
- `dotnet build TimeItSharp.sln -c Release --no-restore`
- `dotnet test TimeItSharp.sln -c Release --no-build --no-restore` — 25 passed
- `dotnet list TimeItSharp.sln package --vulnerable --include-transitive --no-restore` — no vulnerable packages
- Trimmed net8.0 sample publish — passed without trim warnings
- Fluent sample execution across net6.0, net7.0, net8.0, net9.0, and net10.0 — passed
- `dotnet pack` plus local `dotnet tool install` — passed; tool reports version 0.4.8
- `git diff --check` — clean

The only build warnings are the expected NETSDK1138 notices for the retained net6.0/net7.0 compatibility targets. Datadog profiler attachment remains environment/platform dependent; the macOS validation host cannot attach the profiler.

### CODE REVIEW

File: src/TimeItSharp.Common/Exporters/ConsoleExporter.cs [hunk 1]
```diff
@@ -388,7 +423,10 @@ static void GenerateDistributionChart(Dictionary<string, ScenarioResult> dataSer
         var scaledDataSeriesDict = new Dictionary<string, List<double>>();
         foreach (var kvp in dataSeriesDict)
         {
-            var scaledData = kvp.Value.Durations.Select(ns => ns / scale).ToList();
+            var scaledData = kvp.Value.Durations
+                .Where(double.IsFinite)
```
Comment (tonyredondo) [inline src/TimeItSharp.Common/Exporters/ConsoleExporter.cs:504]: Why this where?

File: src/TimeItSharp.Common/Exporters/ConsoleExporter.cs [hunk 2]
```diff
@@ -342,17 +348,46 @@ public void Export(TimeitResult results)
...
         {
             Console.WriteLine("No data available to generate the distribution chart.");
             return;
         }
 
         // Combine all durations from all series to find the overall minimum and maximum
-        var allDataNanoseconds = dataSeriesDict.Values.SelectMany(series => series.Durations).ToList();
+        var allDataNanoseconds = dataSeriesDict.Values
+            .SelectMany(series => series.Durations)
+            .Where(double.IsFinite)
```
Comment (tonyredondo) [inline src/TimeItSharp.Common/Exporters/ConsoleExporter.cs:454]: why do we need this Where

File: src/TimeItSharp.Common/Exporters/DatadogExporter.cs [hunk 3]
```diff
@@ -35,6 +33,13 @@ public DatadogExporter()
     public void Initialize(InitOptions options)
     {
         _options = options;
+        if (!Enabled)
+        {
+            return;
+        }
+
+        Environment.SetEnvironmentVariable("DD_CIVISIBILITY_LOGS_ENABLED", "true");
+        _testSession ??= TestSession.InternalGetOrCreate(Environment.CommandLine, Environment.CurrentDirectory, "time-it");
```
Comment (tonyredondo) [inline src/TimeItSharp.Common/Exporters/DatadogExporter.cs]: I'm worried that this might be too late for initialisation.

File: src/TimeItSharp.Common/Exporters/DatadogExporter.cs [hunk 4]
```diff
@@ -80,7 +106,7 @@ public void Export(TimeitResult results)
                     catch (Exception ex)
                     {
                         AnsiConsole.MarkupLine("[red]Error exporting to datadog:[/]");
-                        AnsiConsole.WriteException(ex);
+                        AnsiConsole.WriteLine(ex.ToString());
```
Comment (tonyredondo) [inline src/TimeItSharp.Common/Exporters/DatadogExporter.cs]: why this change? we have a method to write exceptions!

File: src/TimeItSharp.Common/Exporters/DatadogExporter.cs [hunk 5]
```diff
@@ -119,7 +148,8 @@ public void Export(TimeitResult results)
                 test.SetTag("benchmark.duration.outliers_count", scenarioResult.Outliers?.Count ?? 0);
 
                 // Add metrics
-                if (scenarioResult.MetricsData.TryGetValue("process.time_to_start_ms", out var timeToStart))
+                if (scenarioResult.MetricsData.TryGetValue("process.time_to_start_ms", out var timeToStart) &&
+                    timeToStart.Count > 0)
```
Comment (tonyredondo) [inline src/TimeItSharp.Common/Exporters/DatadogExporter.cs]: with this kind of if we are dropping the full tag in the span, and that's a breaking change

File: src/TimeItSharp.Common/Exporters/DatadogExporter.cs [hunk 6]
```diff
@@ -106,11 +132,14 @@ public void Export(TimeitResult results)
                     RuntimeName = FrameworkDescription.Instance.Name,
                 });
 
-                // Add duration benchmark data
-                test.AddBenchmarkData(
-                    BenchmarkMeasureType.Duration,
-                    "Duration of a run",
-                    BenchmarkDiscreteStats.GetFrom(scenarioResult.Durations.ToArray()));
+                // Add duration benchmark data only when samples were collected.
+                if (scenarioResult.Durations.Count > 0)
```
Comment (tonyredondo) [inline src/TimeItSharp.Common/Exporters/DatadogExporter.cs]: with this kind of if we are dropping the full tag in the span, and that's a breaking change

File: src/TimeItSharp.Common/Exporters/DatadogExporter.cs [hunk 7]
```diff
@@ -228,18 +265,26 @@ public void Export(TimeitResult results)
         {
             errors = true;
             AnsiConsole.MarkupLine("[red]Error exporting to datadog:[/]");
-            AnsiConsole.WriteException(ex);
+            AnsiConsole.WriteLine(ex.ToString());
```
Comment (tonyredondo) [inline src/TimeItSharp.Common/Exporters/DatadogExporter.cs]: again

File: src/TimeItSharp.Common/Services/DatadogProfilerService.cs [hunk 8]
```diff
@@ -331,13 +329,14 @@ static bool GetProfilerPaths(string monitoringHome, ref string? profiler32Path,
             }
 
             if (!File.Exists(profiler64Path) ||
-                (profiler32Path is not null && !File.Exists(profiler64Path)) ||
-                !File.Exists(loaderConfig))
+                (profiler32Path is not null && !File.Exists(profiler32Path)) ||
+                !File.Exists(loaderConfig) ||
+                (ldPreload is not null && !File.Exists(ldPreload)))
```
Comment (tonyredondo) [inline src/TimeItSharp.Common/Services/DatadogProfilerService.cs]: que es esto y por que lo agregas?

File: src/TimeItSharp.Common/ScenarioProcessor.cs [hunk 9]
```diff
@@ -278,13 +268,16 @@ await RunScenarioAsync(repeat.Count, index, scenario, TimeItPhase.ExtraRun, fals
...
+                scenario.ParentService = null;
                 return null;
             }
 
             AnsiConsole.MarkupLine("    Duration: {0}", watch.Elapsed.ToDurationString());
         }
         
         scenario.ParentService = null;
+        var scenarioEnd = DateTime.UtcNow;
+        var scenarioDuration = scenarioEnd - start;
```
Comment (tonyredondo) [inline src/TimeItSharp.Common/ScenarioProcessor.cs]: por que has cambiado el valor del Watch al de un datetime, el stopwatch es mas preciso para duración

File: src/TimeItSharp.Common/ScenarioProcessor.cs [hunk 10]
```diff
@@ -456,8 +457,8 @@ ScenarioResult CreateScenarioResult(AssertResponse response)
                 Metrics = metricsStats,
                 MetricsData = metricsData,
                 Start = start,
-                End = start + watch.Elapsed,
-                Duration = watch.Elapsed,
+                End = scenarioEnd,
+                Duration = scenarioDuration,
```
Comment (tonyredondo) [inline src/TimeItSharp.Common/ScenarioProcessor.cs:620]: por precisión prefiero el Stopwatch

File: src/TimeItSharp.Common/ScenarioProcessor.cs [hunk 11]
```diff
@@ -634,7 +645,7 @@ private async Task<DataPoint> RunCommandAsync(int index, Scenario scenario, Time
         var timeoutCmdArguments = scenario.Timeout.ProcessArguments ?? string.Empty;
 
         var cmdEnvironmentVariables = new Dictionary<string, string?>();
-        foreach (DictionaryEntry osEnv in EnvironmentVariables)
+        foreach (DictionaryEntry osEnv in Environment.GetEnvironmentVariables())
```
Comment (tonyredondo) [inline src/TimeItSharp.Common/ScenarioProcessor.cs]: por que necesitamos esto otra vez? y como se hacia antes?

File: src/TimeItSharp.Common/TimeItSharp.Common.csproj [hunk 12]
```diff
@@ -24,7 +25,13 @@
...
   <ItemGroup>
     <PackageReference Include="CliWrap" Version="3.10.0" />
-    <PackageReference Include="Datadog.Trace.BenchmarkDotNet" Version="2.61.0" />
+    <PackageReference Include="Datadog.Trace.BenchmarkDotNet" Version="2.61.0">
+      <!-- Datadog.Trace.Bundle supplies the newer, complete profiler asset set. -->
+      <ExcludeAssets>contentFiles</ExcludeAssets>
+    </PackageReference>
+    <!-- Bundle includes the matching Datadog.Trace.dll and profiler/native assets.
+         A direct Datadog.Trace reference only supplies Datadog.Trace.Manual.dll. -->
+    <PackageReference Include="Datadog.Trace.Bundle" Version="3.43.0" />
```
Comment (tonyredondo) [inline src/TimeItSharp.Common/TimeItSharp.Common.csproj]: no mezcles la versión 3 con la 2 de ningún paquete de datadog por favor, prefiero que nos quedemos en el ultimo del 2 y ya

File: src/TimeItSharp.Common/Utils.cs [hunk 13]
```diff
@@ -152,6 +168,23 @@ public static bool IsBimodal(Span<double> data, out int peakCount, int binCount
...
+        }
+
+        var normalizedName = name.Replace("_", string.Empty, StringComparison.OrdinalIgnoreCase)
+            .ToUpperInvariant();
+        return normalizedName.Contains("PASSWORD", StringComparison.Ordinal) ||
+               normalizedName.Contains("SECRET", StringComparison.Ordinal) ||
+               normalizedName.Contains("TOKEN", StringComparison.Ordinal) ||
+               normalizedName.Contains("APIKEY", StringComparison.Ordinal) ||
+               normalizedName.Contains("PRIVATEKEY", StringComparison.Ordinal) ||
+               normalizedName.Contains("AUTH", StringComparison.Ordinal);
```
Comment (tonyredondo) [inline src/TimeItSharp.Common/Utils.cs]: agrega también DD_API_KEY y DD_APP_KEY y API_KEY, APP_KEY

File: src/TimeItSharp.Common/TimeItSharp.Common.csproj [hunk 14]
```diff
@@ -24,7 +25,13 @@
 
   <ItemGroup>
     <PackageReference Include="CliWrap" Version="3.10.0" />
-    <PackageReference Include="Datadog.Trace.BenchmarkDotNet" Version="2.61.0" />
+    <PackageReference Include="Datadog.Trace.BenchmarkDotNet" Version="2.61.0">
+      <!-- Datadog.Trace.Bundle supplies the newer, complete profiler asset set. -->
+      <ExcludeAssets>contentFiles</ExcludeAssets>
```
Comment (tonyredondo) [inline src/TimeItSharp.Common/TimeItSharp.Common.csproj]: sin assets no tienes profiler ni tracer

File: src/TimeItSharp.Common/Exporters/ConsoleExporter.cs (hunk 1 above)
Comment (tonyredondo) [reply src/TimeItSharp.Common/Exporters/ConsoleExporter.cs:504]: Good catch. This filter protects the per-series scaling step: `NaN`/`±Infinity` would otherwise propagate into the scaled values and bin indexes and can break chart generation. The final code filters them in the detached result sanitizer and defensively again at the console sink, with malformed custom-result coverage.

File: src/TimeItSharp.Common/Exporters/ConsoleExporter.cs (hunk 2 above)
Comment (tonyredondo) [reply src/TimeItSharp.Common/Exporters/ConsoleExporter.cs:454]: We need this filter before calculating the global min/max and scale. A single non-finite duration would poison the range for every scenario and can produce invalid bins. The final implementation keeps only finite samples and returns a clear no-data result when none remain.

File: src/TimeItSharp.Common/Exporters/DatadogExporter.cs (hunk 3 above)
Comment (tonyredondo) [reply src/TimeItSharp.Common/Exporters/DatadogExporter.cs]: Agreed that initialization timing matters. This is now fixed: enabled exporters are initialized before `BeforeAll`, scenario callbacks, and target execution. The Datadog session/module are created there, while profiler variables are applied by `DatadogProfilerService` during `OnExecutionStart`, before the child is launched. I also removed the process-wide `Environment.SetEnvironmentVariable` mutation from the exporter so it cannot leak into later runs.

File: src/TimeItSharp.Common/Exporters/DatadogExporter.cs (hunk 4 above)
Comment (tonyredondo) [reply src/TimeItSharp.Common/Exporters/DatadogExporter.cs]: Agreed. The final code uses `AnsiConsole.WriteException(Utils.SanitizeException(...))` again. The temporary `WriteLine(ex.ToString())` change was removed; sanitization now happens before calling the existing exception renderer.

File: src/TimeItSharp.Common/Exporters/DatadogExporter.cs (hunk 5 above)
Comment (tonyredondo) [reply src/TimeItSharp.Common/Exporters/DatadogExporter.cs]: Agreed; dropping the measure changes the Datadog span schema. The `Count > 0` gate was removed. When `process.time_to_start_ms` exists we always add the `ApplicationLaunch` benchmark measure, including an empty stats set, while filtering only non-finite samples.

File: src/TimeItSharp.Common/Exporters/DatadogExporter.cs (hunk 6 above)
Comment (tonyredondo) [reply src/TimeItSharp.Common/Exporters/DatadogExporter.cs]: Agreed. The duration benchmark measure is now always added, even for an empty/failed run, so the existing tag/measure contract is preserved. We only discard non-finite samples before building `BenchmarkDiscreteStats`.

File: src/TimeItSharp.Common/Exporters/DatadogExporter.cs (hunk 7 above)
Comment (tonyredondo) [reply src/TimeItSharp.Common/Exporters/DatadogExporter.cs]: Fixed here as well. This path now calls `AnsiConsole.WriteException(Utils.SanitizeException(...))`; raw `ex.ToString()` output is no longer written.

File: src/TimeItSharp.Common/Services/DatadogProfilerService.cs (hunk 8 above)
Comment (tonyredondo) [reply src/TimeItSharp.Common/Services/DatadogProfilerService.cs]: This check was originally added for the proposed v3 API-wrapper/`LD_PRELOAD` path. I removed that design completely. The final profiler contract is v2.61 only and does not use `LD_PRELOAD` or API-wrapper assets; it validates the exact private 2.61 asset set by path, version, file type, ancestry, and SHA-256.

File: src/TimeItSharp.Common/ScenarioProcessor.cs (hunk 9 above)
Comment (tonyredondo) [reply src/TimeItSharp.Common/ScenarioProcessor.cs]: Agreed. Duration is now taken from `scenarioStopwatch.Elapsed`, which is monotonic and more precise. `DateTime` is used only for the timestamp: `scenarioEnd = start + scenarioDuration`.

File: src/TimeItSharp.Common/ScenarioProcessor.cs (hunk 10 above)
Comment (tonyredondo) [reply src/TimeItSharp.Common/ScenarioProcessor.cs:620]: Agreed and restored. `ScenarioResult.Duration` comes from `Stopwatch.Elapsed`; `End` is derived as `start + duration`, so wall-clock changes cannot affect the measured duration.

File: src/TimeItSharp.Common/ScenarioProcessor.cs (hunk 11 above)
Comment (tonyredondo) [reply src/TimeItSharp.Common/ScenarioProcessor.cs]: Previously the child inherited the process environment implicitly through CliWrap. We need an explicit dictionary to overlay scenario variables plus startup-hook/profiler variables and to control `PATH`. The final implementation captures the host environment once at `RunAsync` entry, clones that snapshot per command, and then applies overlays; it no longer rereads/mutates the live host environment on every iteration.

File: src/TimeItSharp.Common/TimeItSharp.Common.csproj (hunk 12 above)
Comment (tonyredondo) [reply src/TimeItSharp.Common/TimeItSharp.Common.csproj]: Agreed. All v3 package/assets were removed. The final package pins `Datadog.Trace.BenchmarkDotNet` and `Datadog.Trace` to exact `[2.61.0]`, keeps a private isolated v2 home, and explicitly rejects v3/mixed/API-wrapper content. The consumer verifier also restores alongside Bundle v3.50 and proves the two homes remain isolated.

File: src/TimeItSharp.Common/Utils.cs (hunk 13 above)
Comment (tonyredondo) [reply src/TimeItSharp.Common/Utils.cs]: Added. Sensitive-name normalization removes separators, so `DD_API_KEY`, `DD_APP_KEY`, `API_KEY`, and `APP_KEY` are covered by the `APIKEY`/`APPKEY` aliases. The same policy is used by console, JSON, Datadog, callback output, and exception sanitization, with regression tests.

File: src/TimeItSharp.Common/TimeItSharp.Common.csproj (hunk 14 above)
Comment (tonyredondo) [reply src/TimeItSharp.Common/TimeItSharp.Common.csproj]: Agreed. The final package includes a complete isolated v2.61 profiler home: 16 BenchmarkDotNet native/config assets plus the four matching managed `Datadog.Trace.dll` files. All 20 files are copied for direct and transitive consumers and validated byte-for-byte/SHA-256 in package, build, publish, wrapper, and v3-collision tests. No v3 or API-wrapper assets are used.
